### PR TITLE
Allow procmacro exported items to be renamed.

### LINKED ISSUE #1753: Allow specifying primary constructor
Use case:

I have an existing Rust library (or a library I wish to have an idiomatic Rust interface). It has a struct 

```rust
#[derive(uniffi::Object)]
pub struct MyLibraryKey {
    bytes: [u8; 512]
}
#[uniffi:export]
impl MyLibraryKey {
    #[uniffi::constructor]
    pub fn new(bytes: [u8; 512]) -> Self {
        Self {
            bytes
        }
    }
}
```

Currently the `primary_constructor` is inferred from it being named `new`. Foreign languages don't support fixed size arrays, but I'd like to keep my ergonomic Rust API surface so I do something like:

```rust
impl MyLibraryKey {
    pub fn new(bytes: [u8; 512]) -> Self {
        Self {
            bytes
        }
    }
}
#[uniffi::export]
impl MyLibrary {
    #[uniffi::constructor]
    pub fn new_ffi(bytes: Vec<u8>) -> Result<Self, String> {
        Ok(Self {
            bytes: bytes.try_into().map_err(|_| "wrong byte length")?
        })
    }
}
```

Expectation:
I want to be able to tell `uniffi` that `new_ffi` is the primary constructor so downstream languages use it in their native constructors (ie `MyLibraryKey(b"deadbeef")`), but Rust keeps its ergonomic interface as well.

Potential Solution:
Something like `#[uniffi::constructor(is_primary=true)]` would work, but doing that in a way that doesn't break current usage that expects `new` to be primary may be tricky. Putting `#[cfg(feature = "ffi")]` and `#[cfg(not(feature = "ffi"))]` around each `impl` would be a workaround but as things get more complex could get extremely unwieldy.

### PR DESCRIPTION
**Pull request by mhammond**
Fixes #1753

While #1753 talks just about the constructor, I noticed that we already had the capability of using different names for exported items but just lacked a way to express that.

### CODE REVIEW

File: docs/manual/src/proc_macro/index.md
Comment (jplatte) [inline docs/manual/src/proc_macro/index.md:146]:
```suggestion
#[uniffi::export(name = something)]
```

File: docs/manual/src/proc_macro/index.md
Comment (jplatte) [inline docs/manual/src/proc_macro/index.md:158]: Why would the c'tor name not be part of `constructor`s arguments?

File: docs/manual/src/proc_macro/index.md
Comment (jplatte) [reply docs/manual/src/proc_macro/index.md:146]: Also I wonder if a bare word should be used here, or a string. If it's a bare word, using a keyword in that position will likely result in it being highlighted, but other than that I can't think of any practical consequences.

File: uniffi_macros/src/export/attributes.rs
Comment (jplatte) [inline uniffi_macros/src/export/attributes.rs:97]: `.parse::<Ident>()` will forbid Rust keywords in the input, which seems wrong, I could see one primary use for this feature being when you wanna use a Rust keyword as the foreign language name. `Ident::parse_any` gets around that.

File: fixtures/proc-macro/src/lib.rs
Comment (jplatte) [inline fixtures/proc-macro/src/lib.rs:325]: How about another dummy attribute like constructor (just to have a place to attach method attributes) instead of using one-item impls?


```suggestion
#[uniffi::export]
impl Renamed {
    #[uniffi::method(name = func)]
    fn renamed_func(&self) -> bool {
        true
    }
}
```

File: docs/manual/src/proc_macro/index.md
Comment (mhammond) [reply docs/manual/src/proc_macro/index.md:146]: I was going back and forward here. My initial thought was that an ident is closer to the intent and would help avoid obviously incorrect values (eg, with a space), but the more I look at it, the more I think it probably should just be a string - having an ident there makes my brain pause as I scan over it, wondering where this `something` came from. Making it a string makes it clearer we are just assigning a literal value and as you mention, causes less surprising syntax highlighting..

File: uniffi_macros/src/export/attributes.rs
Comment (mhammond) [reply uniffi_macros/src/export/attributes.rs:97]: As above, I'll change to a string, but I think we [already support Rust keywords](https://github.com/mozilla/uniffi-rs/tree/main/fixtures/keywords/rust/src) as foreign language names.

Comment (jplatte) [reply uniffi_macros/src/export/attributes.rs]: Yeah, supporting it elsewhere is why it should not cause a parse error here either. But moot point anyways, as you noted.

File: uniffi_macros/src/export/attributes.rs
Comment (mhammond) [inline uniffi_macros/src/export/attributes.rs:307]: do we really need this - maybe this can just be left unchecked and this becomes a noop?

File: docs/manual/src/proc_macro/index.md
Comment (Sajjon) [inline docs/manual/src/proc_macro/index.md:146]: This is nice! Any plans to support renaming of `args`? That would be really fantastic! In the realm of Swift, we typically use prepositions a lot. so `from, `to`, `by` etc is the preferred label, e.g. [see Swift standard library `stride:from:to:by`](https://developer.apple.com/documentation/swift/stride(from:to:by:)). But since Rust does not have the notion of seperated "labels" (externally shown) and argname (internal to the body of the function) like Swift does, as Rust devs it is strange to use `from`, `to`, `by` as arg names.

So it would be amazing of `uniffi::export(name` could support renames of args!

Bikeshedding syntax:
```rust
#[uniffi::export(fn_name = "stride", arg_names = "from,to,by")]
fn do_stride(start: usize, end: usize, stride_length: usize) {
}
```

Would that be interesting to add support for?

Would it be possible?

**Review by jplatte — COMMENTED** [review]
Haven't had a look in a while but I guess we can merge and improve later, right? If you want a proper second review though, just say so and I'll try to find some time for it.

**Review by jplatte — APPROVED** [review]
Review above was supposed to be an approval.

**Pull request merged**
